content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md

```diff
@@ -0,0 +1,181 @@
+
+A new kube-proxy parameter will be added to specify the IPVS load balancing algorithm, with the parameter being `--ipvs-scheduler`. If it’s not configured, then round-robin (rr) is the default value.
+
+- rr: round-robin
+- lc: least connection
+- dh: destination hashing
+- sh: source hashing
+- sed: shortest expected delay
+- nq: never queue
+
+In the future, we can implement Service specific scheduler (potentially via annotation), which has higher priority and overwrites the value.
+
+**Parameter: --cleanup-ipvs** Similar to the --cleanup-iptables parameter, if true, cleanup IPVS configuration and IPTables rules that are created in IPVS mode.
+
+**Parameter: --ipvs-sync-period** Maximum interval of how often IPVS rules are refreshed (e.g. '5s', '1m'). Must be greater than 0.
+
+**Parameter: --ipvs-min-sync-period** Minimum interval of how often the IPVS rules are refreshed (e.g. '5s', '1m'). Must be greater than 0.
+
+**Parameter: --ipvs-exclude-cidrs**  A comma-separated list of CIDR's which the IPVS proxier should not touch when cleaning up IPVS rules because IPVS proxier can't distinguish kube-proxy created ipvs rules from user original ipvs rules. If you are using IPVS proxier with your own IPVS rules in the environment, this parameter should be specified, otherwise your original rule will be cleaned.
+
+### Design Considerations
+
+#### IPVS Service Network Topology
+
+When creating a ClusterIP type Service, IPVS proxier will do the following three things:
+
+* Make sure a dummy interface exists in the node, default to kube-ipvs0
+* Bind Service IP addresses to the dummy interface
+* Create IPVS virtual servers for each Service IP address respectively
+
+Here comes an example:
+
+```
+# kubectl describe svc nginx-service
+Name:			nginx-service
+...
+Type:			ClusterIP
+IP:			    10.102.128.4
+Port:			http	3080/TCP
+Endpoints:		10.244.0.235:8080,10.244.1.237:8080
+Session Affinity:	None
+
+# ip addr
+...
+73: kube-ipvs0: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN qlen 1000
+    link/ether 1a:ce:f5:5f:c1:4d brd ff:ff:ff:ff:ff:ff
+    inet 10.102.128.4/32 scope global kube-ipvs0
+       valid_lft forever preferred_lft forever
+
+# ipvsadm -ln
+IP Virtual Server version 1.2.1 (size=4096)
+Prot LocalAddress:Port Scheduler Flags
+  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn     
+TCP  10.102.128.4:3080 rr
+  -> 10.244.0.235:8080            Masq    1      0          0         
+  -> 10.244.1.237:8080            Masq    1      0          0   
+```
+
+Please note that the relationship between a Kubernetes Service and IPVS virtual servers is `1:N`. Consider a Kubernetes Service that has more than one IP addresses, for example, an External IP type Service has two IP addresses(ClusterIP and External IP). Then the IPVS proxier will create 2 IPVS virtual servers - one for Cluster IP and another one for External IP. The relationship between a Kubernetes Endpoint(each IP+Port pair) and an IPVS virtual server is `1:1`.
```

Comment:
spaces before `(each IP+Port pair)` and `(ClusterIP and External IP)`.

i would reword:
> Consider a Kubernetes Service that has more than one IP addresses, for example, an External IP type Service has two IP addresses(ClusterIP and External IP).

to:

> For example, consider a Kubernetes Service that has more than one IP address. An External IP type Service has two IP addresses - ClusterIP and External IP.